extensions/sql-database-projects/src/test/project.test.ts

```diff
@@ -743,6 +745,55 @@ describe('Project: add SQLCMD Variables', function (): void {
 	});
 });
 
+describe('Project: properties', function (): void {
+	before(async function (): Promise<void> {
+		await baselines.loadBaselines();
+	});
+
+	it('Should read target database version', async function (): Promise<void> {
+		projFilePath = await testUtils.createTestSqlProjFile(baselines.openProjectFileBaseline);
+		const project = await Project.openProject(projFilePath);
+
+		should(project.getProjectTargetVersion()).equal('150');
+	});
+
+	it('Should throw on missing target database version', async function (): Promise<void> {
+		projFilePath = await testUtils.createTestSqlProjFile(baselines.sqlProjectMissingVersionBaseline);
+		const project = await Project.openProject(projFilePath);
+
+		should(() => project.getProjectTargetVersion()).throw("Invalid DSP in .sqlproj file");
```

Comment:
use the error strings in constants.ts

Reply:
Hm, why? These are tests, not a production code. In tests we want to validate that we get the expected value. If the code is changing, then expected value should be updated. If you use a production constant in tests, then whenever the constant is updated to return an incorrect value - tests will not catch it.

If it helps to understand, then what you are suggesting is basically to write a test like `should(add(2, 2)).equal(add(2, 2))`. I don't think that is how tests are supposed to be written. Expected value simply cannot come from the production code that you are testing.

Reply:
I think there's a difference between the two depending on what the intention of the check is.  By using the error string, you're saying that the important thing is that the error gets thrown, not so much the specific string of the error; if the string gets updated, the test will - by design - still pass as long as the logic leading to throwing an error is intact.  By using a string literal, you're putting greater emphasis on the _contents_ of the string which, IMO, makes more sense when the string is processed in some way (as in your `add(2, 2)` example).  Don't think it matters too much one way or the other for this particular check.

Reply:
I think your explanation would make sense, if the ask was to remove the string altogether and only keep `.throw()` to validate that there is an error thrown, but the ask was to use production code in the expected value, which is a rather strange practice, to say the least.

Reply:
Even then, when using the production constant, it still checks that the correct error is thrown (over a different error).  The only thing it doesn't protect from is someone accidentally checking in a mistake to that particular const (say, a typo or some copy pasta), but in doing so, it allows for single-line updates to the error message instead of requiring an edit in every test that confirms it.  Different, but not necessarily wrong or inferior, IMO.  Either way, feel free to leave as is since you obviously feel strongly about this.

Reply:
> The only thing it doesn't protect from is someone accidentally checking in a mistake to that particular const (say, a typo or some copy pasta)

Unfortunately, that is not the only thing. In simple words, having the value that you expect in the test definition checks that the implementation in production code is correct. For instance, in this particular case, it also verifies that `localize` call performs as it should (or what matters for this particular codebase, as you don't want to test `localize` call - that strings are localized). The "constant" has an implementation and by using it here as the expected value you are basically claiming that production code does what it should and essentially should not be tested, which as I said is rather strange practice.

> it allows for single-line updates to the error message instead of requiring an edit in every test that confirms it

I think we should get on the same page here first. Writing and maintaining tests does imply an additional effort, sometimes this effort is times more, than the actual implementation. If additional effort incurred by tests is your concern than you should probably revisit the decision to have them at first place. Now, the solution to this particular problem is to introduce constant values in tests themselves, but never to reference production code as expected value.

> not necessarily wrong or inferior

I'd argue that it is in fact inferior, as it does not clearly explain the expectation. For instance, I'm currently having an issue with this PR, where expected behavior was never clearly set in the tests, and was updated in production code. This significantly diminishes the value of the tests at first place, as they can no longer serve as a reference point for the correct implementation. This is simply another example of that - if tomorrow somebody replaces the `localize` call with hard-coded string - your "different" implementation of the tests will pass, but in reality it should fail when tested for any non-"en-US" locale (which has 0% coverage today, by the way).

> feel free to leave as is since you obviously feel strongly about this.

Not only I plan to leave this as-is, but I'd also suggest your team to revisit basic coding practices.